hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractInvocationFuture.java

```diff
@@ -17,71 +17,99 @@
 import static java.util.concurrent.locks.LockSupport.unpark;
 
 /**
  * Custom implementation of {@link java.util.concurrent.CompletableFuture}.
- * <p>
- * The long term goal is that this whole class can be ripped out and replaced
- * by {@link java.util.concurrent.CompletableFuture} from the JDK. So we need
- * to be very careful with adding more functionality to this class because
- * the more we add, the more
- * difficult the replacement will be.
- * <p>
- * TODO:
- * - thread value protection
- *
  * @param <V>
  */
-@SuppressWarnings("Since15")
 @SuppressFBWarnings(value = "DLS_DEAD_STORE_OF_CLASS_LITERAL", justification = "Recommended way to prevent classloading bug")
-public abstract class AbstractInvocationFuture<V> implements InternalCompletableFuture<V> {
+@SuppressWarnings({"checkstyle:methodcount", "checkstyle:ClassDataAbstractionCoupling", "checkstyle:ClassFanOutComplexity"})
+public abstract class AbstractInvocationFuture<V> extends InternalCompletableFuture<V> {
+
+    static final Object UNRESOLVED = new Object();
+    private static final Lookup LOOKUP = MethodHandles.publicLookup();
+    // new Throwable(String message, Throwable cause)
+    private static final MethodType MT_INIT_STRING_THROWABLE = MethodType.methodType(void.class, String.class, Throwable.class);
+    // new Throwable(Throwable cause)
+    private static final MethodType MT_INIT_THROWABLE = MethodType.methodType(void.class, Throwable.class);
+    // new Throwable(String message)
+    private static final MethodType MT_INIT_STRING = MethodType.methodType(void.class, String.class);
+
+    private static final AtomicReferenceFieldUpdater<AbstractInvocationFuture, Object> STATE =
+            newUpdater(AbstractInvocationFuture.class, Object.class, "state");
 
-    static final Object VOID = "VOID";
+    // Default executor for async callbacks: ForkJoinPool.commonPool() or a thread-per-task executor when
+    // the common pool does not support parallelism
+    private static final Executor DEFAULT_ASYNC_EXECUTOR;
 
     // reduce the risk of rare disastrous classloading in first call to
     // LockSupport.park: https://bugs.openjdk.java.net/browse/JDK-8074773
     static {
         @SuppressWarnings("unused")
         Class<?> ensureLoaded = LockSupport.class;
-    }
 
-    private static final AtomicReferenceFieldUpdater<AbstractInvocationFuture, Object> STATE =
-            newUpdater(AbstractInvocationFuture.class, Object.class, "state");
+        Executor asyncExecutor;
+        if (ForkJoinPool.getCommonPoolParallelism() > 1) {
+            asyncExecutor = ForkJoinPool.commonPool();
+        } else {
+            asyncExecutor = command -> new Thread(command).start();
+        }
+        DEFAULT_ASYNC_EXECUTOR = asyncExecutor;
+    }
 
-    protected final Executor defaultExecutor;
     protected final ILogger logger;
 
+    // todo since we are using FJP.commonPool by default (under conditions), remove this field
```

Comment:
todo, just not to forget about it

Reply:
Done; after removing this field and the lambdas from `get()`/`join()`/`joinInternal()` the allocation overhead of `CompletableFuture` vs current `master`'s `ICompletableFuture` in a simple `IAtomicLong#getAndIncrement` benchmark is ~2% (previously was ~6%).